### ISSUE TITLE
UseHttpMethodOverride not working in .NET 8.0 RC1

### ISSUE THREAD

**Issue by andrewdavey** [dotnet/aspnetcore#50815]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I'm trying to upgrade a .NET 6 web application to .NET 8 RC1. I have `app.UseHttpMethodOverride()` to allow client to post with the `X-Http-Method-Override` header.

I'm getting 405 (Method Not Allowed) responses from the endpoints that are mapping non-POST methods. My code is not getting called at all. It's like the pipeline is failing to use the modified HTTP method to route correctly.

### Expected Behavior

My endpoint for a non-POST HTTP method should be called.

### Steps To Reproduce

Minimal reproduction:

```c#
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();
app.UseHttpMethodOverride();
app.MapDelete("/", () => "delete");
app.Run();
```

Run that, then make a request e.g. with powershell:

```powershell
Invoke-WebRequest -Uri http://localhost:5070/ -Method POST -Headers @{ 'X-Http-Method-Override' = 'DELETE' }
```

### .NET Version

8.0.100-rc.1.23463.5

**Comment by captainsafia** [comment]
Thanks for reporting this issue, @andrewdavey!

I believe this is related to the fact that the routing middleware is enabled implicitly in the WebApplication and runs _before_ the HttpOverrides middleware. Can you see if invoking the middlewares explicitly in the correct order works for you? It works on my end.

```csharp
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

app.UseHttpMethodOverride();
app.UseRouting();

app.MapDelete("/", () => "delete");
app.Run();
```

**Comment by andrewdavey** [comment]
Thanks for the quick response. Inserting `UseRouting()` does fix my initial example. However, my real application is using controllers. I'm still getting a 405 status response when requesting controller actions that use attribute mapping.

Example:

```c#
using Microsoft.AspNetCore.Mvc;

var builder = WebApplication.CreateBuilder(args);
builder.Services.AddControllers();

var app = builder.Build();

app.UseHttpMethodOverride();
app.UseRouting();
app.MapControllers();
app.Run();

public class TestController : Controller
{
    [HttpDelete("/")]
    public string Delete() => "delete";
}
```

**Comment by viliml** [comment]
`builder.Services.AddControllers();` registers authorization services: https://github.com/dotnet/aspnetcore/blob/46562b1435bf111a7425b40f507b157b42a016a4/src/Mvc/Mvc/src/MvcServiceCollectionExtensions.cs#L132-L138
When authorization services are registered, the authorization middleware is inserted: https://github.com/dotnet/aspnetcore/blob/46562b1435bf111a7425b40f507b157b42a016a4/src/DefaultBuilder/src/WebApplicationBuilder.cs#L447-L452
***Before*** all the user-defined middlewares: https://github.com/dotnet/aspnetcore/blob/46562b1435bf111a7425b40f507b157b42a016a4/src/DefaultBuilder/src/WebApplicationBuilder.cs#L465-L467
The authorization middleware inserts a new instance of routing middleware immediately after itself as long as global routing is already defined: https://github.com/dotnet/aspnetcore/blob/46562b1435bf111a7425b40f507b157b42a016a4/src/Security/Authorization/Policy/src/AuthorizationAppBuilderExtensions.cs#L36-L43
However, global routing is always defined: https://github.com/dotnet/aspnetcore/blob/46562b1435bf111a7425b40f507b157b42a016a4/src/DefaultBuilder/src/WebApplicationBuilder.cs#L412-L413
The authorization rerouting middleware runs before the HTTP method override middleware and sends the request to 405, and the subsequent user-defined routing middleware doesn't dare question its predecessor's decision: https://github.com/dotnet/aspnetcore/blob/46562b1435bf111a7425b40f507b157b42a016a4/src/Http/Routing/src/EndpointRoutingMiddleware.cs#L61-L67
Removing authorization fixes the problem in this particular scenario:
```cs
using Microsoft.AspNetCore.Mvc;

var builder = WebApplication.CreateBuilder(args);
builder.Services.AddMvcCore();

var app = builder.Build();

app.UseHttpMethodOverride();
app.UseRouting();
app.MapControllers();
app.Run();

public class TestController : Controller
{
    [HttpDelete("/")]
    public string Delete() => "delete";
}
```
To properly fix the underlying problem, the HTTP method override middleware needs to reroute the request after changing the method. I implemented that [here](https://github.com/viliml/aspnetcore/tree/patch-2) (3aff15b8dd5501fe499f2d64c80cde661000e733) and verified it works, but I daren't open a pull request since I'm not sure I did it in the best possible way.

**Comment by viliml** [comment]
Would also be fixed by #50864

**Comment by captainsafia** [comment]
@viliml Spot on!

@andrewdavey Explicitly registering all the middlewares in the correct order should do the trick while we wait to roll out RC2.
